README.md
<span align="center">

# @switchbot/homebridge-switchbot-ble

[![npm version](https://badgen.net/npm/v/@switchbot/homebridge-switchbot-ble)](https://www.npmjs.com/package/@switchbot/homebridge-switchbot-ble)
[![npm downloads](https://badgen.net/npm/dt/@switchbot/homebridge-switchbot-ble)](https://www.npmjs.com/package/@switchbot/homebridge-switchbot-ble)

</span>

The @switchbot/homebridge-switchbot-ble is a nodejs module , and also a homebridge plug-in that directly controls SwitchBot products via BLE.

Now supports:
 * [SwitchBot (Bot)](https://www.switch-bot.com/products/switchbot-bot)
 * [SwitchBot Curtain (Curtain)](https://www.switch-bot.com/products/switchbot-curtain)
 * [SwitchBot Thermometer & Hygrometer (Meter)](https://www.switch-bot.com/products/switchbot-meter)

## Installation
### Installing bluetooth libraries
##### Ubuntu, Debian, Raspbian
```sh
sudo apt-get install bluetooth bluez libbluetooth-dev libudev-dev
```
See the document of the [@abandonware/noble](https://github.com/abandonware/noble#readme) for other operating systems details.

### Installing package
```sh
sudo npm install -g @switchbot/homebridge-switchbot-ble
```
You can also install it on the homebridge plugins page.
![homebridge-plugins-search](image/homebridge-plugins-search.png)

## Configuration
Add this to your homebridge `config.json` file
```json
"platforms": [
    {
        "platform": "SwitchBotPlatform",
        "name": "SwitchBotPlatform"
    }
]
```
If it work, add devices.
```json
"platforms": [
    {
        "platform": "SwitchBotPlatform",
        "name": "SwitchBotPlatform",
        "devices": [
            {
                "type": "bot",
                "name": "Bot 2c",
                "bleMac": "d4:cc:43:97:ae:2c"
            },
            {
                "type": "bot",
                "name": "Bot 51",
                "bleMac": "e7:4d:36:cf:9e:51",
                "scanDuration": 2000
            },
            {
                "type": "curtain",
                "name": "Curtain 11",
                "bleMac": "ec:58:c5:d0:01:11",
                "scanDuration": 2000,
                "reverseDir": false,
                "moveTime": 2000
            },
            {
                "type": "meter",
                "name": "Meter 84",
                "bleMac": "d0:d5:64:14:43:84",
                "scanDuration": 5000,
                "scanInterval": 60000
            }
        ]
    }
]
```

**Requird Settings**
* `devices` - SwitchBot devices list.
* `type` - Device type. Currently supports `bot`, `curtain` and `meter`.
* `name` - Device name.
* `bleMac` - Device mac address. You can find it in App settings.

**Optional Settings**
* `scanDuration` - Scan timeout. BLE Central must first scan the advertising. Default is `1000`(unit: ms). Longer time to ensure device discovery but slower response.
* `reverseDir` - Set to `true` to exchange the "opened" and "closed" directions of Curtain after calibration. Default is `false`. So you can swap the directions without recalibration.
* `moveTime` - Set the actual running time of Curtian, which is also the time to switch from running state to opened/closeed state in the Home APP. Default is `2000`(unit: ms).
* `scanInterval` - Scan interval. Currently indicates the cycle of updating the temperature and humidity value of Meter. Default is `60000`(unit: ms).

Please note that:

1. The device `name` cannot be the same.
2. Restart HomeBridge after updating the config file.
3. The "opened" and "closed" directions of HomeKit are the same as those in Curtain and SwitchBot APP, but the percentage definitions are different. "Closed" is 0% in HomeKit but 100% in Curtain and APP.

## Community

* [SwitchBot (Official website)](https://www.switch-bot.com/)
* [Facebook @SwitchBotRobot](https://www.facebook.com/SwitchBotRobot/) 
* [Twitter @SwitchBot](https://twitter.com/switchbot) 


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. This project uses [Semantic Versioning](https://semver.org/)

## [Version 1.2.7](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-12-22)

### Changes

- Fixed issues from `v1.2.6`.

## [Version 1.2.6](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-12-21)

### Changes

- Fixed issue with `config.schema.json` not being packaged in `v1.2.4`.
- Update dependencies.

## [Version 1.2.5](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-12-21)

### Changes

- Fixed issue with `config.schema.json` not being packaged in `v1.2.4`.
- Update Dependencies.

## [Version 1.2.4](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-12-21)

### Changes

- Support Homebridge config setting UI. (Thanks to [@donavanbecker](https://github.com/SwitchBot/homebridge-switchbot-ble/pull/24))
- Fixed TypeError: this.config.devices is not iterable. (Thanks to [@donavanbecker](https://github.com/SwitchBot/homebridge-switchbot-ble/pull/23))

## [Version 1.2.3](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-11-30)

### Changes

- Fixed the incorrect state of Bot in Home App when control fails.
- Modify log messages. (Thanks to [@dnicolson](https://github.com/SwitchBot/homebridge-switchbot-ble/pull/15))

## [Version 1.2.2](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-11-19)

### Changes

- Fixed Curtain state will not be updated in the Home App.

## [Version 1.2.1](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-11-12)

### Changes

- Get temperature and humidity from Meter advertising periodically.

## [Version 1.2.0](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-11-06)

### Changes

- Fixed HomeKit control direction reversal issue.
- Add option for inverted curtain position. (Thanks to [@whatUwant](https://github.com/SwitchBot/homebridge-switchbot-ble/pull/4))

## [Version 1.1.0](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-10-29)

### Changes

- Add support for Curtain.

## [Version 1.0.1](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-10-22)

### Changes

- First public release, supports Bot.


src/bot-accessory.ts
/* eslint-disable brace-style */
/* eslint-disable prefer-const */
/* eslint-disable eqeqeq */
/* eslint-disable @typescript-eslint/no-unused-vars */
import {
  AccessoryPlugin,
  CharacteristicGetCallback,
  CharacteristicSetCallback,
  CharacteristicValue,
  HAP,
  Logging,
  Service,
  CharacteristicEventTypes,
} from "homebridge";
import { rejects } from "assert";

export class Bot implements AccessoryPlugin {

  private readonly log: Logging;
  private readonly bleMac: string;
  private readonly scanDuration: number;

  private switchOn = false;
  private runTimer!: NodeJS.Timeout;

  // This property must be existent!!
  name: string;

  private readonly botService: Service;
  private readonly informationService: Service;

  constructor(hap: HAP, log: Logging, name: string, bleMac: string, scanDuration: number) {
    this.log = log;
    this.name = name;
    this.bleMac = bleMac;
    this.scanDuration = scanDuration;

    this.botService = new hap.Service.Switch(name);
    this.botService.getCharacteristic(hap.Characteristic.On)
      .on(CharacteristicEventTypes.GET, (callback: CharacteristicGetCallback) => {
        log.info("Current state of Bot was returned: " + (this.switchOn ? "ON" : "OFF"));
        callback(undefined, this.switchOn);
      })
      .on(CharacteristicEventTypes.SET, (value: CharacteristicValue, callback: CharacteristicSetCallback) => {
        const targetState = value as boolean;
        clearTimeout(this.runTimer);
        if (targetState === this.switchOn) {
          log.info("Target state of Bot has not changed: " + (this.switchOn ? "ON" : "OFF"));
          this.botService?.getCharacteristic(hap.Characteristic.On).updateValue(this.switchOn);
          callback();
        }
        // Target state has been changed.
        log.info("Target state of Bot setting: " + (targetState ? "ON" : "OFF"));
        const SwitchBot = require('node-switchbot');
        const switchbot = new SwitchBot();
        switchbot.discover({ duration: this.scanDuration, model: 'H', quick: false }).then((device_list: any) => {
          log.info('Scan done.');
          let targetDevice: any = null;
          for (let device of device_list) {
            // log.info(device.modelName, device.address);
            if (device.address == this.bleMac) {
              targetDevice = device;
              break;
            }
          }
          if (!targetDevice) {
            log.info('No device was found.');
            return new Promise((resolve, reject) => {
              reject(new Error('No device was found.'));
            });
          }
          else {
            log.info(targetDevice.modelName + ' (' + targetDevice.address + ') was found.');
            // Set event handers
            targetDevice.onconnect = () => {
              // log.info('Connected.');
            };
            targetDevice.ondisconnect = () => {
              // log.info('Disconnected.');
            };
            log.info('Bot is running...');
            if (targetState) {
              return targetDevice.turnOn();
            }
            else {
              return targetDevice.turnOff();
            }
          }
        }).then(() => {
          log.info('Done.');
          this.switchOn = targetState;
          this.runTimer = setTimeout(() => {
            this.botService?.getCharacteristic(hap.Characteristic.On).updateValue(this.switchOn);
          }, 500);
          log.info("Bot state has been set to: " + (this.switchOn ? "ON" : "OFF"));
          callback();
        }).catch((error: any) => {
          log.error(error);
          this.runTimer = setTimeout(() => {
            this.botService?.getCharacteristic(hap.Characteristic.On).updateValue(this.switchOn);
          }, 500);
          log.info("Bot state failed to be set to: " + (targetState ? "ON" : "OFF"));
          callback();
        });
      });

    this.informationService = new hap.Service.AccessoryInformation()
      .setCharacteristic(hap.Characteristic.Manufacturer, "SwitchBot")
      .setCharacteristic(hap.Characteristic.Model, "SWITCHBOT-S1")
      .setCharacteristic(hap.Characteristic.SerialNumber, this.bleMac);

    log.info("Bot '%s' created!", name);
  }

  /*
   * This method is optional to implement. It is called when HomeKit ask to identify the accessory.
   * Typical this only ever happens at the pairing process.
   */
  identify(): void {
    this.log.info("Identify!");
  }

  /*
   * This method is called directly after creation of this instance.
   * It should return all services which should be added to the accessory.
   */
  getServices(): Service[] {
    return [
      this.informationService,
      this.botService,
    ];
  }

}


src/meter-accessory.ts
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/no-inferrable-types */
import {
  AccessoryPlugin,
  API,
  CharacteristicGetCallback,
  CharacteristicSetCallback,
  CharacteristicValue,
  HAP,
  Logging,
  Service,
  CharacteristicEventTypes,
} from "homebridge";
import { rejects } from "assert";

export class Meter implements AccessoryPlugin {

  private readonly log: Logging;
  private readonly bleMac: string;
  private readonly scanDuration: number;
  private readonly scanInterval: number;

  private temperature: number = 0;
  private humidity: number = 0;

  // This property must be existent!!
  name: string;
  displayName: string;

  private readonly temperatureSercice: Service;
  private readonly humidityService: Service;
  private readonly informationService: Service;
  private loggingService;

  constructor(homebridge: API, hap: HAP, log: Logging, name: string, bleMac: string, scanDuration: number, scanInterval: number) {
    this.log = log;
    this.name = name;
    this.displayName = name;
    this.bleMac = bleMac;
    this.scanDuration = scanDuration;
    this.scanInterval = scanInterval;

    var FakeGatoHistoryService = require('fakegato-history')(homebridge);
    this.loggingService = new FakeGatoHistoryService("weather", this, {
    disableTimer: true,
    storage:'googleDrive',
    forlder:'FakeGato',
    keyPath:'/home/pi/fakegato/'
    });

    this.temperatureSercice = new hap.Service.TemperatureSensor(name);
    this.temperatureSercice.getCharacteristic(hap.Characteristic.CurrentTemperature)
      .on(CharacteristicEventTypes.GET, (callback: CharacteristicGetCallback) => {
        log.info(name + " current temperature: " + this.temperature + "\u2103");
        callback(undefined, (this.temperature < 0) ? 0 : (this.temperature > 100 ? 100 : this.temperature));
      })
      .on(CharacteristicEventTypes.SET, (value: CharacteristicValue, callback: CharacteristicSetCallback) => {
        log.info("The temperature of the Meter can't be set!");
        callback();
      });

    this.humidityService = new hap.Service.HumiditySensor(name);
    this.humidityService.getCharacteristic(hap.Characteristic.CurrentRelativeHumidity)
      .on(CharacteristicEventTypes.GET, (callback: CharacteristicGetCallback) => {
        log.info(name + " current humidity: " + this.humidity + "%");
        callback(undefined, this.humidity);
      })
      .on(CharacteristicEventTypes.SET, (value: CharacteristicValue, callback: CharacteristicSetCallback) => {
        log.info("The humidity of the Meter can't be set!");
        callback();
      });

    this.informationService = new hap.Service.AccessoryInformation()
      .setCharacteristic(hap.Characteristic.Manufacturer, "SwitchBot")
      .setCharacteristic(hap.Characteristic.Model, "SWITCHBOT-METERTH-S1")
      .setCharacteristic(hap.Characteristic.SerialNumber, this.bleMac);

    log.info(name, "scanDuration:" + this.scanDuration.toString() + "ms", "scanInterval:" + this.scanInterval.toString() + "ms");

    const Switchbot = require('node-switchbot');
    const switchbot = new Switchbot();

    switchbot.onadvertisement = (ad: any) => {
      // log.info(JSON.stringify(ad, null, '  '));
      // log.info("Temperature:", ad.serviceData.temperature.c);
      // log.info("Humidity:", ad.serviceData.humidity);
      this.temperature = ad.serviceData.temperature.c;
      this.humidity = ad.serviceData.humidity;
      this.loggingService.addEntry({time: Math.round(new Date().valueOf() / 1000), temp: this.temperature, humidity: this.humidity});

    };

    switchbot.startScan({
      id: bleMac,
    }).then(() => {
      return switchbot.wait(this.scanDuration);
    }).then(() => {
      switchbot.stopScan();
    }).catch((error: any) => {
      log.error(error);
    });

    setInterval(() => {
      // log.info("Start scan " + name + "(" + bleMac + ")");
      switchbot.startScan({
        // mode: 'T',
        id: bleMac,
      }).then(() => {
        return switchbot.wait(this.scanDuration);
      }).then(() => {
        switchbot.stopScan();
        // log.info("Stop scan " + name + "(" + bleMac + ")");
      }).catch((error: any) => {
        log.error(error);
      });
    }, this.scanInterval);
  }

  /*
   * This method is optional to implement. It is called when HomeKit ask to identify the accessory.
   * Typical this only ever happens at the pairing process.
   */
  identify(): void {
    this.log.info("Identify!");
  }

  /*
   * This method is called directly after creation of this instance.
   * It should return all services which should be added to the accessory.
   */
  getServices(): Service[] {
    return [
      this.informationService,
      this.temperatureSercice,
      this.humidityService,
      this.loggingService,
    ];
  }

}


src/curtain-accessory.ts
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable eqeqeq */
/* eslint-disable prefer-const */
/* eslint-disable brace-style */
import {
  AccessoryPlugin,
  CharacteristicGetCallback,
  CharacteristicSetCallback,
  CharacteristicValue,
  HAP,
  Logging,
  Service,
  CharacteristicEventTypes,
} from "homebridge";
import { rejects } from "assert";

export class Curtain implements AccessoryPlugin {

  private readonly log: Logging;
  private readonly bleMac: string;
  private readonly scanDuration: number;
  private readonly reverseDir: boolean;
  private readonly moveTime: number;

  private currentPosition = 0;
  private targetPosition = 0;
  private positionState = 0;
  private moveTimer!: NodeJS.Timeout;

  // This property must be existent!!
  name: string;

  private readonly curtainService: Service;
  private readonly informationService: Service;

  constructor(hap: HAP, log: Logging, name: string, bleMac: string, scanDuration: number, reverseDir: boolean, moveTime: number) {
    this.log = log;
    this.name = name;
    this.bleMac = bleMac;
    this.scanDuration = scanDuration;
    this.reverseDir = reverseDir;
    this.moveTime = moveTime;

    this.positionState = hap.Characteristic.PositionState.STOPPED;

    this.curtainService = new hap.Service.WindowCovering(name);
    this.curtainService.getCharacteristic(hap.Characteristic.CurrentPosition)
      .on(CharacteristicEventTypes.GET, (callback: CharacteristicGetCallback) => {
        log.info("Current position of Curtain was returned: " + this.currentPosition + "%");
        callback(undefined, this.currentPosition);
      });

    this.curtainService.getCharacteristic(hap.Characteristic.TargetPosition)
      .on(CharacteristicEventTypes.GET, (callback: CharacteristicGetCallback) => {
        log.info("Target position of Curtain was returned: " + this.targetPosition + "%");
        callback(undefined, this.targetPosition);
      })
      .on(CharacteristicEventTypes.SET, (value: CharacteristicValue, callback: CharacteristicSetCallback) => {
        this.targetPosition = value as number;
        log.info("Target position of Curtain setting: " + this.targetPosition + "%");
        clearTimeout(this.moveTimer);
        if (this.targetPosition > this.currentPosition) {
          this.positionState = hap.Characteristic.PositionState.INCREASING;
        }
        else if (this.targetPosition < this.currentPosition) {
          this.positionState = hap.Characteristic.PositionState.DECREASING;
        }
        else {
          this.positionState = hap.Characteristic.PositionState.STOPPED;
        }

        if (this.positionState === hap.Characteristic.PositionState.STOPPED) {
          this.curtainService?.getCharacteristic(hap.Characteristic.TargetPosition).updateValue(this.targetPosition);
          this.curtainService?.getCharacteristic(hap.Characteristic.CurrentPosition).updateValue(this.currentPosition);
          this.curtainService?.getCharacteristic(hap.Characteristic.PositionState).updateValue(this.positionState);
          callback();
        }
        else {
          const SwitchBot = require('node-switchbot');
          const switchbot = new SwitchBot();
          switchbot.discover({ duration: this.scanDuration, model: 'c', quick: false }).then((device_list: any) => {
            log.info('Scan done.');
            let targetDevice: any = null;
            for (let device of device_list) {
              log.info(device.modelName, device.address);
              if (device.address == this.bleMac) {
                targetDevice = device;
                break;
              }
            }
            if (!targetDevice) {
              log.info(name + ' (' + bleMac + ') was not found.');
              return new Promise((resolve, reject) => {
                reject(new Error('No target device was found.'));
              });
            }
            else {
              log.info(targetDevice.modelName + ' (' + targetDevice.address + ') was found.');
              // Set event handers
              targetDevice.onconnect = () => {
                // log.info('Connected.');
              };
              targetDevice.ondisconnect = () => {
                // log.info('Disconnected.');
              };
              log.info('Curtain is moving...');
              /**opend - 0% in HomeKit, 100% in Curtain device.
               * closed - 100% in HomeKit, 0% in Curtain device.
               * To keep the status synchronized, convert the percentage of homekit to the percentage of Curtain.
               */
              let covertToDevicePosition = 0;
              if (!reverseDir) {
                covertToDevicePosition = 100 - this.targetPosition;
              }
              else {
                log.info('Reverse the "opened" and "closed" directions!');
                covertToDevicePosition = this.targetPosition;
              }
              return targetDevice.runToPos(covertToDevicePosition);
            }
          }).then(() => {
            log.info('Done.');
            log.info("Target position of Curtain has been set to: " + this.targetPosition + "%");
            this.moveTimer = setTimeout(() => {
              // log.info("setTimeout", this.positionState.toString(), this.currentPosition.toString(), this.targetPosition.toString());
              this.currentPosition = this.targetPosition;
              this.positionState = hap.Characteristic.PositionState.STOPPED;
              // this.curtainService?.getCharacteristic(hap.Characteristic.TargetPosition).updateValue(this.targetPosition);
              this.curtainService?.getCharacteristic(hap.Characteristic.CurrentPosition).updateValue(this.currentPosition);
              this.curtainService?.getCharacteristic(hap.Characteristic.PositionState).updateValue(this.positionState);
            }, this.moveTime);
            callback();
          }).catch((error: any) => {
            log.error(error);
            this.moveTimer = setTimeout(() => {
              this.targetPosition = this.currentPosition;
              this.positionState = hap.Characteristic.PositionState.STOPPED;
              this.curtainService?.getCharacteristic(hap.Characteristic.TargetPosition).updateValue(this.targetPosition);
              // this.curtainService?.getCharacteristic(hap.Characteristic.CurrentPosition).updateValue(this.currentPosition);
              this.curtainService?.getCharacteristic(hap.Characteristic.PositionState).updateValue(this.positionState);
            }, 1000);
            log.info("Target position of Curtain failed to be set to: " + this.targetPosition + "%");
            callback();
          });
        }
      });

    this.curtainService.getCharacteristic(hap.Characteristic.PositionState)
      .on(CharacteristicEventTypes.GET, (callback: CharacteristicGetCallback) => {
        log.info("The position state of Curtain was returned: " + this.positionState);
        callback(undefined, this.positionState);
      });

    this.informationService = new hap.Service.AccessoryInformation()
      .setCharacteristic(hap.Characteristic.Manufacturer, "SwitchBot")
      .setCharacteristic(hap.Characteristic.Model, "SWITCHBOT-CURTAIN-W0701600")
      .setCharacteristic(hap.Characteristic.SerialNumber, this.bleMac);

    log.info("Curtain '%s' created!", name);
  }

  /*
   * This method is optional to implement. It is called when HomeKit ask to identify the accessory.
   * Typical this only ever happens at the pairing process.
   */
  identify(): void {
    this.log.info("Identify!");
  }

  /*
   * This method is called directly after creation of this instance.
   * It should return all services which should be added to the accessory.
   */
  getServices(): Service[] {
    return [
      this.informationService,
      this.curtainService,
    ];
  }

}


src/switchbot-platform.ts
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable no-var */
/* eslint-disable no-case-declarations */
/* eslint-disable prefer-const */
import { AccessoryPlugin, 
  API, 
  HAP, 
  Logging, 
  PlatformConfig, 
  StaticPlatformPlugin, 
} from "homebridge";
import { Bot } from "./bot-accessory";
import { Curtain } from "./curtain-accessory";
import { Meter } from "./meter-accessory";
import { off } from "process";

const PLATFORM_NAME = "SwitchBotPlatform";

/*
 * IMPORTANT NOTICE
 *
 * One thing you need to take care of is, that you never ever ever import anything directly from the "homebridge" module (or the "hap-nodejs" module).
 * The above import block may seem like, that we do exactly that, but actually those imports are only used for types and interfaces
 * and will disappear once the code is compiled to Javascript.
 * In fact you can check that by running `npm run build` and opening the compiled Javascript file in the `dist` folder.
 * You will notice that the file does not contain a `... = require("homebridge");` statement anywhere in the code.
 *
 * The contents of the above import statement MUST ONLY be used for type annotation or accessing things like CONST ENUMS,
 * which is a special case as they get replaced by the actual value and do not remain as a reference in the compiled code.
 * Meaning normal enums are bad, const enums can be used.
 *
 * You MUST NOT import anything else which remains as a reference in the code, as this will result in
 * a `... = require("homebridge");` to be compiled into the final Javascript code.
 * This typically leads to unexpected behavior at runtime, as in many cases it won't be able to find the module
 * or will import another instance of homebridge causing collisions.
 *
 * To mitigate this the {@link API | Homebridge API} exposes the whole suite of HAP-NodeJS inside the `hap` property
 * of the api object, which can be acquired for example in the initializer function. This reference can be stored
 * like this for example and used to access all exported variables and classes from HAP-NodeJS.
 */
let hap: HAP;
let homebridge: API;

export = (api: API) => {
  hap = api.hap;
  homebridge = api;

  api.registerPlatform(PLATFORM_NAME, SwitchBotPlatform);
};

class SwitchBotPlatform implements StaticPlatformPlugin {

  private readonly log: Logging;
  private readonly config: PlatformConfig;

  constructor(log: Logging, config: PlatformConfig, api: API) {
    this.log = log;

    // probably parse config or something here
    this.config = config;

    log.info("SwitchBot platform finished initializing!");
  }

  /*
   * This method is called to retrieve all accessories exposed by the platform.
   * The Platform can delay the response my invoking the callback at a later time,
   * it will delay the bridge startup though, so keep it to a minimum.
   * The set of exposed accessories CANNOT change over the lifetime of the plugin!
   */
  accessories(callback: (foundAccessories: AccessoryPlugin[]) => void): void {
    let deviceList = [];
    if (this.config.devices) {
      for (var device of this.config.devices) {
        // this.log.info(device.type);
        // this.log.info(device.name);
        // this.log.info(device.bleMac);
        // this.log.info(device.scanDuration, typeof device.scanDuration);
        let scanDuration: number = device.scanDuration || 1000;
        switch (device.type) {
          case 'bot':
            deviceList.push(new Bot(hap, this.log, device.name, device.bleMac.toLowerCase(), scanDuration));
            break;
          case 'curtain':
            const reverseDir: boolean = device.reverseDir || false;
            const moveTime: number = device.moveTime || 2000;
            deviceList.push(new Curtain(hap, this.log, device.name, device.bleMac.toLowerCase(), scanDuration, reverseDir, moveTime));
            break;
          case 'meter':
            let scanInterval: number = device.scanInterval || 60000;
            if (scanInterval < scanDuration) {
              scanInterval = scanDuration + 1000;
            }
            deviceList.push(new Meter(homebridge, hap, this.log, device.name, device.bleMac.toLowerCase(), scanDuration, scanInterval));
            break;
          default:
            break;
        }
      }
    } else {
      this.log.error('No Device Set In Config');
    }
    this.log("Device amount:", deviceList.length.toString());
    callback(deviceList);
  }
}
